File: server/managers/rsa.js
const NodeRSA = require('node-rsa');

exports("GenerateKeys", (cb) => {
    if (GetInvokingResource()) {
        const key = new NodeRSA({b: 1024});

        cb(key.exportKey("public"), key.exportKey("private"))
    }
})

exports("Decrypt", (privatekey, encrypted, cb) => {
    if (GetInvokingResource()) {
        // Key importation
        const key = new NodeRSA();
        key.importKey(privatekey, "pkcs1-private")

        // Decryption
        const decrypted = key.decrypt(encrypted, "utf8")

        // Return
        cb(decrypted)
    }
})

exports("Encrypt", (publickey, text, cb) => {
    if (GetInvokingResource()) {
        // Key importation
        const key = new NodeRSA();
        key.importKey(publickey, "pkcs8-public")

        // Encryption
        const encrypted = key.encrypt(text, "base64")

        // Return
        cb(encrypted)
    }
})

File: client/managers/rsa.lua
local BlacklistedEncrypted = {}

RequestDataFromNui = function(message, type)
    local event = nil
    local p = promise:new()

    event = RegisterNetEvent("__cfx_nui:"..type, function(data)
        p:resolve(data)
    end)

    SendNUIMessage(message)

    local data = Citizen.Await(p)
    RemoveEventHandler(event)
    collectgarbage("collect")

    return data
end

RegisterNuiCallbackType("GenerateKey")
RegisterNuiCallbackType("Decrypted")
RegisterNuiCallbackType("Encrypted")

GenerateKeys = function()
    local data = RequestDataFromNui({type = "generate"}, "GenerateKey")
    return data.publicKey, data.privateKey
end
Encrypt = function(public, text)
    local data = RequestDataFromNui({type = "encrypt", public = public, text = text}, "Encrypted")

    return data.encrypted
end
Decrypt = function(private, encrypted)
    local data = RequestDataFromNui({type = "decrypt", private = private, encrypted = encrypted}, "Decrypted")

    return data.decrypted
end

exports("GenerateKeys", GenerateKeys)
exports("Encrypt", Encrypt)
exports("Decrypt", Decrypt)